### PR TITLE
Add explicit delete permission to link shares

### LINKED ISSUE #24868: Unable to delete files inside folders shared by link
### Steps to reproduce
1. On server A create a new folder (aka folder1)
2. On server A create a new file inside folder1 (aka file1)
3. On server A share by link folder1 (allow edit checked)
4. Paste link on browser, click on Add to your owncloud and fill ownCloud field with server B url 
5. On server B accept shared folder
6. On server B navigate inside folder1 and check available options for file1

### Expected behaviour
Shared by link should have the same behaviour than shared option

Shared by link:
screen shot 2016-05-27 at 09 25 53

Shared:

screen shot 2016-05-27 at 09 40 45

### Actual behaviour

As user who accepts shared items, we are not able to remove files inside folder shared by link. On 9.0.2 it works fine, shows the same options.

### Server configuration
**Operating system**:
Ubuntu 14.04.1 LTS
**Web server:**
Apache2
**Database:**
Server version: 5.5.46-0ubuntu0.14.04.2 (Ubuntu)
**PHP version:**
PHP 5.5.9-1ubuntu4.14 (cli) (built: Oct 28 2015 01:34:46
**ownCloud version:** (see ownCloud admin page)
ownCloud Enterprise Edition 9.1.0 beta 1 (testing)

**The content of config/config.php:**

```
<?php
$CONFIG = array (
  'updatechecker' => false,
  'supportedDatabases' => 
  array (
    0 => 'mysql',
    1 => 'pgsql',
    2 => 'oci',
  ),
  'instanceid' => 'oco48z2585bc',
  'passwordsalt' => 'ZZZZ',
  'secret' => 'XXXXX',
  'trusted_domains' => 
  array (
    0 => 'localhost',
    1 => 'XXXXXX',
  ),
  'datadirectory' => '/opt/owncloud/data',
  'overwrite.cli.url' => 'XXXX',
  'dbtype' => 'mysql',
  'version' => '9.1.0.5',
  'dbname' => 'owncloud',
  'dbhost' => 'localhost',
  'dbtableprefix' => '',
  'dbuser' => 'oc_admin',
  'dbpassword' => '1xgukCfhszo+kYlKpQN8ZK0JCmrvHo',
  'logtimezone' => 'UTC',
  'installed' => true,
  'license-key' => 'demo-20160625-80e046c4d642dfa0-a9a2295f',
);

```
**Are you using encryption:**

No

@rullzer

### PR DESCRIPTION
**Pull request by PVince81**
Link shares always allowed deletion, however internally the permissions
were stored as 7 which lacked delete permissions. This created an
inconsistency in the Webdav permissions.

This fix makes sure we include delete permissions in the share
permissions, which now become 15.

In case a client is still passing 7 for legacy reasons, it gets
converted automatically to 15.

Fixes https://github.com/owncloud/core/issues/24868

TODOs:
- [x] JS code sends 15 instead of 7
- [x] OCS API adjusted to accept both 7 and 15 but normalizes to 15
- [x] Repair step to convert 7 to 15 for all link shares in DB

@rullzer I settled on normalizing to make sure we don't break existing clients that might be using the "permissions" attribute when updating link shares. I do hope they use "publicUpload" instead.

### CODE REVIEW

**Comment by PVince81** [comment]
Repair step added.

Please review @owncloud/sharing @icewind1991 @mcastroSG

**Comment by PVince81** [comment]
- [x] TEST: create from web UI
- [x] TEST: create from OCS API
- [x] TEST: update from web UI
- [x] TEST: update from OCS API with publicUpload attr
- [x] TEST: update from OCS API with permissions attr
- [ ] TEST: repair step

**Comment by PVince81** [comment]
I tested the repair step by upgrading from v9.1.0beta2 to this branch and have a read-write link share. The permission got converted to 15.

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by PVince81** [comment]
Second review please @rullzer @owncloud/sharepoint @owncloud/filesystem @ChristophWurst @georgehrke

**Comment by PVince81** [comment]
@jvillafanez

File: apps/files_sharing/lib/API/Share20OCS.php [hunk 1]
```diff
@@ -602,12 +603,21 @@ public function updateShare($id) {
 
 			if ($newPermissions !== null &&
 				$newPermissions !== \OCP\Constants::PERMISSION_READ &&
-				$newPermissions !== (\OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE)) {
+				// legacy
+				$newPermissions !== (\OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE) &&
+				// correct
+				$newPermissions !== (\OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE | \OCP\Constants::PERMISSION_DELETE)
+			) {
```
Comment (jvillafanez) [inline apps/files_sharing/lib/API/Share20OCS.php]: I'd rewrite the condition as:
```
$permission_r = \OCP\Constants::PERMISSION_READ;
$permission_cru = \OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE;
$permission_crud = \OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE | \OCP\Constants::PERMISSION_DELETE;
if (!in_array($new_permissions, [null, $permission_r, $permission_cru, $permission_crud], true)) {
.....
}
```

And something similar for the other cases

**Comment by jvillafanez** [comment]
I'd only try to rewrite the constants and the conditions.

There are quite a bunch of `\OCP\Constants::PERMISSION_READ | \OCP\Constants::PERMISSION_CREATE | \OCP\Constants::PERMISSION_UPDATE`. I don't know if there is a proper place to set this as a constant, or at least as a static variable, so it's easier and shorter to read

**Comment by jvillafanez** [comment]
The rest of the code looks good

**Comment by PVince81** [comment]
> at least as a static variable, so it's easier and shorter to read

@jvillafanez I thought about this but from what I remember you can't define static constants that use other static constants, or it might not work properly.

**Comment by mcastroSG** [comment]
Have made some test and seems to be working perfectly. Using a server with this patch applied we are able to delete files inside shared by link folders.

**Comment by jvillafanez** [comment]
@PVince81 Yes, you're right. I've checked the docs, and you can't define constants or variables with expressions (at least when you're defining the constant or variable in PHP < 5.6).

What about using an utility function somewhere? Does it make sense?

**Comment by georgehrke** [comment]
👍

**Comment by PVince81** [comment]
> What about using an utility function somewhere? Does it make sense?

Or local attributes. But for now I don't want to refactor too much of this, the focus was fixing the issue.

**Comment by jvillafanez** [comment]
ok, fair enough 👍

**Pull request merged**